File: Problems/2342. Max Sum of a Pair With Equal Sum of Digits/Solution.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
private:
    int sumDigits(int num) {
        int sum = 0;
        while (num) {
            sum += num % 10;
            num /= 10;
        }
        return sum;
    }
public:
    int maximumSum(vector<int>& nums) {
        vector<int> sameDigit(82, -1);
        int sum, maxSum = -1;
        for(int num : nums) {
            sum = sumDigits(num);
            if (sameDigit[sum] != -1) {
                maxSum = max(maxSum, sameDigit[sum] + num);
                sameDigit[sum] = max(sameDigit[sum], num);
            }
            else {
                sameDigit[sum] = num;
            }
        }
        return maxSum;
    }
};

File: Problems/2342. Max Sum of a Pair With Equal Sum of Digits/Solution.rs
impl Solution {
    pub fn maximum_sum(nums: Vec<i32>) -> i32 {
        let mut same_digit: Vec<i32> = vec![-1; 82];
        let mut max_sum: i32 = -1;

        for num in &nums {
            let (mut val, mut sum): (i32, usize) = (*num, 0);
            while val > 0 {
                sum += (val % 10) as usize;
                val /= 10;
            }
            
            let tmp_max = same_digit[sum];
            if tmp_max > 0 {
                max_sum = max_sum.max(*num + tmp_max);
            }
            same_digit[sum] = tmp_max.max(*num);
        }

        return max_sum;
    }
}